### DISCUSSION TITLE
re-using a guard which evaluates context

### DISCUSSION THREAD

**Discussion by lougreenwood** [statelyai/xstate#1376] [General]
Hello, I'm just beginning my journey with Xstate, and sketching out an initial design for a machine, but I'm hitting some un-expected behaviour which I wondered someone could help me with.

Here's my state chart: https://xstate.js.org/viz/?gist=3a0a4e19b4c4788c00ef3c451ea2d94c

I have a context defined and a"transient transition" out of `unknown` to select a `thing`. I tried to re-use the same guard since in my final version of this machine, I will have `n` things, so I didn't want to clutter the guards. The problem is that it seems when re-using a guard, any uses of that guard will return as true, even if the specific called guard is false.

e.g, in the above state chart, I expect that `thingB` would be the selected state, instead both guards evaluate true (green) and `thingA` is the selected stated (presumably) since it's first in the list.
This one works as expected, but I now have to have n guards to match each state...

https://xstate.js.org/viz/?gist=01bc19e9685e9a5e85bf70fa5ae74819

Is this expected behaviour, if so, how would I model this?

**Comment by davidkpiano** [comment]
> Is this expected behaviour

It's not a question of whether this is "expected behavior" -- you modeled the guard so that it is essentially:

```js
return context.thingy === 'thingA' || context.thingy === 'thingB'
```

That is a completely different guard than the guards you have in the second example, and yes, it is expected (in JS, this is not XState-specific) that the function will return `true` for either `'thingA'` or `'thingB'`.

You can parameterize the guard to reduce repetition:

```js
// ...
cond: { type: 'isMatching', thing: 'thingA' },

// ...
cond: { type: 'isMatching', thing: 'thingB' },

guards: {
  isMatching: (ctx, e, { cond }) => {
    return ctx.thingy === cond.thing;
  }
}
```

See [the docs on custom guards](https://xstate.js.org/docs/guides/guards.html#custom-guards) for more info.
```

**Reply by lougreenwood** [reply]
Damn, how did I miss that.... It's late 🤦 - thank you.

Thanks for the help... Alternatively, do you know if it's possible to somehow access the target state which is being evaluated for transition in the guard itself. That would remove the need to repeat the target state in the custom guard param as you described (and what I think I managed to convince myself I was somehow doing when I was just evaluating against strings in my idiot example 😉).
